### PR TITLE
GH-37500: [CI][C++] Disable Dataset and Substrait by default

### LINKED ISSUE #37500: [CI][C++] C++ build script enables Dataset and Substrait by default
### Describe the bug, including details regarding any error messages, version, and platform.

Dataset and Substrait are heavy subcomponents, so they should be disabled in the baseline CI script by default.

### Component(s)

C++, Continuous Integration

### PR DESCRIPTION
**Pull request by pitrou**
### Rationale for this change

Datasets and Substrait are heavy subcomponents (as they have non-trivial dependencies), so should not be enabled unwillingly in all CI builds.

### What changes are included in this PR?

Disable Dataset and Substrait in `ci/scripts/cpp_build.sh`, enable them explicitly in relevant builds.

Disable more subcomponents in CI builds that don't need them, especially "minimal" builds.

### Are these changes tested?

Yes, by existing CI jobs.

### Are there any user-facing changes?

No.

* Closes: #37500

### CODE REVIEW

**Comment by pitrou** [comment]
@github-actions crossbow submit -g cpp -g python

**Comment by pitrou** [comment]
cc @kou

**Comment by pitrou** [comment]
@github-actions crossbow submit -g example

**Comment by pitrou** [comment]
For some reason, the so-called minimal builds are still not minimal at all.

**Comment by pitrou** [comment]
@github-actions crossbow submit test-conda* example*

File: .github/workflows/ruby.yml [hunk 1]
```diff
@@ -121,6 +121,8 @@ jobs:
       ARROW_BUILD_STATIC: OFF
       ARROW_BUILD_TESTS: OFF
       ARROW_BUILD_UTILITIES: OFF
+      ARROW_DATASET: ON
+      ARROW_FLIGHT: ON
```
Comment (kou) [inline .github/workflows/ruby.yml:125]:
```suggestion
```

**Review by kou — APPROVED** [review]
+1

**Pull request merged**